purchases/src/test/java/com/revenuecat/purchases/attributes/SubscriberAttributesDeviceCacheTests.kt

```diff
@@ -163,46 +180,249 @@ class SubscriberAttributesDeviceCacheTests {
+        )
+        val cacheContents = mapOf(
+            "cesar" to expectedAttributes,
+            "pedro" to expectedAttributes
+        )
+        mockLegacyCacheMultipleUsers(cacheContents)
+
+        val allLegacyStoredSubscriberAttributes = underTest.getAllLegacyStoredSubscriberAttributes()
+
+        assertThat(allLegacyStoredSubscriberAttributes.size).isEqualTo(cacheContents.size)
+        allLegacyStoredSubscriberAttributes.forEach { (userID, receivedAttributes) ->
+            expectedAttributes.values.map { it to receivedAttributes[it.key.backendKey] }
+                .forEach { (expected, received) ->
+                    assertThat(received).isNotNull
+                    assertThat(expected).isEqualTo(received)
+                }
+        }
+    }
+
+    @Test
+    fun `Given there are no legacy subscriber attributes`() {
+        val cacheContents = emptyMap<AppUserID, SubscriberAttributeMap>()
+        mockLegacyCacheMultipleUsers(cacheContents)
+
+        val allLegacyStoredSubscriberAttributes = underTest.getAllLegacyStoredSubscriberAttributes()
+
+        assertThat(allLegacyStoredSubscriberAttributes.size).isEqualTo(cacheContents.size)
+    }
+
+    @Test
+    fun `Given there are no legacy subscriber attributes, migration does't do anything`() {
+        mockLegacyCacheMultipleUsers(emptyMap())
+
+        underTest.migrateSubscriberAttributesIfNeeded()
+
+        verify (exactly = 0) { mockEditor.remove(any()) }
+        verify (exactly = 0) { mockEditor.putString(any(), any()) }
+    }
+
+    @Test
+    fun `Given there are legacy subscriber attributes, migration happens`() {
+        val expectedAttributes = mapOf(
+            "tshirtsize" to SubscriberAttribute("tshirtsize", "L", isSynced = true),
+            "age" to SubscriberAttribute("age", "L", isSynced = true)
+        )
+        val legacyCacheContents = mapOf(
+            "cesar" to expectedAttributes,
+            "pedro" to expectedAttributes
+        )
+        legacyCacheContents.keys.forEach {
+            every {
+                mockEditor.remove("com.revenuecat.purchases.$apiKey.subscriberAttributes.$it")
+            } returns mockEditor
+        }
+        mockLegacyCacheMultipleUsers(legacyCacheContents)
+        val cacheContents = mapOf(
+            appUserID to expectedAttributes,
+            "user2" to expectedAttributes
+        )
+        mockNotEmptyCacheMultipleUsers(cacheContents)
+
+        underTest.migrateSubscriberAttributesIfNeeded()
+
+        verify (exactly = 1) {
+            mockEditor.remove("com.revenuecat.purchases.$apiKey.subscriberAttributes.pedro")
+        }
+        verify (exactly = 1) {
+            mockEditor.remove("com.revenuecat.purchases.$apiKey.subscriberAttributes.cesar")
+        }
+        assertCapturedEqualsExpected(mapOf(
+            "cesar" to expectedAttributes,
+            "pedro" to expectedAttributes,
+            appUserID to expectedAttributes,
+            "user2" to expectedAttributes
```

Comment:
I think it might be a good idea to have at least one case where different users have different attributes, to cover against bugs where we're assigning the same attributes to all users

Reply:
we should also have one case with old + new attributes found

Reply:
I updated this test to also include that case and make it more complete